4-WebApp-your-API/4-1-MyOrg/TodoListService/Controllers/TodoListController.cs

```diff
@@ -27,66 +31,161 @@ public TodoListController(IHttpContextAccessor contextAccessor)
         {
+            var owner = GetObjectIdClaim(User);
+
+            if (HasApplicationPermissions(new string[] { "ToDoList.ReadWrite.All" }))
+            {
+                //with such a permission any owner name is accepted from UI
+                owner = todo.Owner;
+            }
+
             int id = TodoStore.Values.OrderByDescending(x => x.Id).FirstOrDefault().Id + 1;
-            Todo todonew = new Todo() { Id = id, Owner = HttpContext.User.Identity.Name, Title = todo.Title };
+            Todo todonew = new Todo() { Id = id, Owner = owner, Title = todo.Title };
             TodoStore.Add(id, todonew);
 
             return Ok(todo);
         }
 
         // PATCH api/values
         [HttpPatch("{id}")]
-        [RequiredScope("ToDoList.Write")]
+        [RequiredScopeOrAppPermission(
+            AcceptedScope = new string[] { "ToDoList.ReadWrite" },
+            AcceptedAppPermission = new string[] { "ToDoList.ReadWrite.All" })]
         public IActionResult Patch(int id, [FromBody] Todo todo)
         {
-            if (id != todo.Id)
+            if (id != todo.Id || !TodoStore.Values.Any(x => x.Id == id))
             {
                 return NotFound();
             }
 
-            if (TodoStore.Values.FirstOrDefault(x => x.Id == id) == null)
+            if (
+                HasDelegatedPermissions(new string[] { "ToDoList.ReadWrite" })
+                && TodoStore.Values.Any(x => x.Id == id && x.Owner == GetObjectIdClaim(User))
+                && todo.Owner == GetObjectIdClaim(User)
+
+                ||
+
+                HasApplicationPermissions(new string[] { "ToDoList.ReadWrite.All" })
+
+                )
             {
-                return NotFound();
+                TodoStore.Remove(id);
+                TodoStore.Add(id, todo);
+
+                return Ok(todo);
             }
 
-            TodoStore.Remove(id);
-            TodoStore.Add(id, todo);
+            throw new ApplicationException("You have insufficient permissions to run this patch operation.");
 
-            return Ok(todo);
+        }
+
+        //check if the permission is inside claims
+        private bool HasApplicationPermissions(string[] permissionsNames)
+        {
+            var rolesClaim = User.Claims.Where(
+              c => c.Type == ClaimConstants.Roles || c.Type == ClaimConstants.Role)
+              .SelectMany(c => c.Value.Split(' '));
+
+            var result = rolesClaim.Any(v => permissionsNames.Any(p => p.Equals(v)));
+
+            return result;
+        }
+
+        //check if the scope is inside claims
+        private bool HasDelegatedPermissions(string[] scopesNames)
+        {
+            var result = (User.FindFirst(ClaimConstants.Scp) ?? User.FindFirst(ClaimConstants.Scope))?
+                .Value.Split(' ').Any(v => scopesNames.Any(s => s.Equals(v)));
+
+            return result ?? false;
+        }
+
+        private string GetObjectIdClaim(ClaimsPrincipal user)
```

Comment:
Not needed. Already exists as an extension method in Id.Web.

Reply:
Yes, Replaced everywhere by _contextAccessor.HttpContext.User.GetObjectId()
Left the method as educational alternative

Reply:
Please check the Idweb one and use it..